riot/riot.py

```diff
@@ -272,26 +267,101 @@ def needs_venv(self) -> bool:
     @property
     def pkg_str(self) -> str:
         """Return pip friendly install string from defined packages."""
+        chain: t.List[VenvInstance] = []
+        current: t.Optional[VenvInstance] = self
+        while current is not None:
+            chain.append(current)
+            current = current.parent
+
+        pkgs: t.Dict[str, str] = {}
+        for inst in chain:
+            pkgs.update(dict(inst.pkgs))
+
         return " ".join(
             [
                 f"'{get_pep_dep(lib, version)}'"
-                for lib, version in self.pkgs
+                for lib, version in pkgs.items()
                 if version is not None
             ]
         )
 
     @property
-    def pythonpath(self) -> str:
-        """Return the expected PYTHONPATH env variable for this Venv.
+    def scriptpath(self):
+        paths = []
+        current = self
+        while current is not None:
+            if current.needs_venv:
+                assert current.target is not None, current
+                paths.append(os.path.join(current.target, "bin"))
+            current = current.parent
+        return ":".join(paths)
 
-        This will include the Interpreter's Python path, and if there
-        are pkgs defined, this venvs Python path appended.
-        """
-        paths = [self.py.pythonpath]
-        if self.needs_venv:
-            paths.extend(get_venv_sitepackages(self.venv_path()))
+    @property
+    def pythonpath(self):
+        paths = []
+        current = self
+        while current is not None:
+            if current.needs_venv:
+                assert current.target is not None, current
+                version = ".".join((str(_) for _ in self.py.version_info()[:2]))
+                paths.append(
+                    os.path.join(
+                        current.target, "lib", f"python{version}", "site-packages"
+                    )
+                )
+            current = current.parent
         return ":".join(paths)
 
+    def prepare(
+        self, env: t.Dict[str, str], py: t.Optional[Interpreter] = None
+    ) -> None:
+        # Propagate the interpreter down the parenting relation
+        self.py = py = py or self.py
+
+        if py is not None:
+            venv_path = self.venv_path
+            if venv_path is None:
+                return
+
+            if self.needs_venv:
+                py_ex = py.path()
+                logger.info(
+                    "Creating virtualenv '%s' with interpreter '%s'.",
+                    self.venv_path,
+                    py_ex,
+                )
+
+                try:
+                    run_cmd(
```

Comment:
I think there is no need to create more venvs other than the base ones with this change, but I think it's best to experiment with this in a follow-up PR.

Reply:
well, that is what we handle with `if self.needs_venv` right? we only need to create additional venvs whenever an intermediary venv has dependencies?

Reply:
I think it is enough to run `pip install` within the base environment because we then specify the location of each dependency based on the "venv" configuration with `--prefix`. We can then decide which packages to actually use in the base venv by adding paths to `PYTHONPATH`. So by running inside the base venv with the right `PYTHONPATH` we should be OK.

Reply:
you are suggesting we do the opposite of what we have today, which is run everything in the base venv, and then add "child"/"other" site packages paths to add the necessary dependencies

compared to today, which we create the child venvs with the extra dependencies, and then add the parent venv site packages to PYTHONPATH to grab all the base dependencies + the library.



writing it out, it doesn't seem like there is actually that much of a difference between the two?

Reply:
nope, I just read the description for #143 and I was reading this comment wrong.

we create the base venv with base library, and then use pip prefix to just segment where we have pip install the unique/extra packages so we skip creating entire child venvs

that sounds good to me.